Add automatic settings for shipping rates

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Implements part of #1117.

This PR adds a new `automatic` setting to the shipping rates (we currently have `manual` and `flat`). This new setting will instruct the backend to automatically sync the Merchant Center shipping settings with WooCommerce shipping settings. It uses the WooCommerce shipping zones to get a good estimate of shipping rates in the defined countries and submits them to Google.

I say "estimate" because the shipping zones can be defined in so many complex ways and the Google shipping settings can only handle a limited scope of what can be defined in Woo. We try to map most of the settings 1:1 but there are many cases where we either have to skip or pick one of many available options because there can't be a 1:1 mapping. For example, defining a shipping rate based on a formula containing variables is not supported in Merchant Center; therefore, we skip those shipping rates at the moment.

I had to do some refactoring to the `MerchantCenterService` and extract a `TargetAudience` service from it because the functionality related to fetching the target countries was required in the `Settings` class and since this class is also used in `MerchantCenterService`, if I had used the methods from `MerchantCenterService` it would've created a dependency loop. Most of the changes in this PR are side effects of refactoring this class.

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Define multiple shipping zones each containing various regions and shipping methods and rates for shipping classes.
2. Send a `POST` request to `wc/gla/mc/settings` and include the following data as JSON body: `{ "shipping_rate":"automatic" }`
3. Confirm that the setting is updated in DB. It's stored in `wp_options` under key: `gla_merchant_center`
4. Send a `POST` request to `wc/gla/mc/settings/sync`
5. Confirm that it's successful
6. Check the shipping rates in your MC account and confirm that they match what you defined in Woo shipping zones


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>

Commit:
Add `TargetAudience` service

This service contains the methods required for fetching the store's target audience.

Commit:
Add method to get rates in the standard format

Commit:
Add unit tests

Commit:
Get suggested rates from shipping zone service

Commit:
Add `automatic` shipping rate type to the API

Commit:
Submit the shipping rates from Woo settings if `automatic` is set

Commit:
Fix unit test

jorgemd24 commented on src/Shipping/ShippingZone.php:
```diff
@@ -78,6 +78,79 @@ public function get_shipping_methods_for_country( string $country_code ): array
+			if ( self::METHOD_PICKUP === $method['id'] ) {
+				continue;
+			}
+
```
❓️

What do you think about having a method called `get_supported_shippings_rates` ( or similar name ) that returns only the methods available?

` $supported_methods = $this->get_supported_shippings_rates( $methods ); `

nima-karimi replied:
The check for pickup method was added temporarily until we figure out if we can include the pickup method in this release. It seems like we can't so I removed it. The methods are already checked in `parse_method` function to see if we support them but just so that we have a more clear place, I've added a check in https://github.com/woocommerce/google-listings-and-ads/pull/1192/commits/ab8bf5552f485eda5318281d0836b4355b03568d.

jorgemd24 replied:
Thanks for the clarification!

jorgemd24 commented on src/Shipping/ShippingZone.php:
```diff
@@ -378,4 +452,26 @@ public static function is_shipping_method_valid( string $method ): bool {
+		$free_shipping_method = array_filter(
+			$methods,
+			function ( $method ) {
+				return self::METHOD_FREE === $method['id'];
```
❓ 

I think we need to add `$method['enable'] === true` otherwise it will return a $free_shipping_method even when is disabled.

nima-karimi replied:
Yes good catch. Added it in https://github.com/woocommerce/google-listings-and-ads/pull/1192/commits/796623752b289c71a8145343e8d40acb8674579b

jorgemd24 replied:
Thanks! 👍

jorgemd24 commented on src/Shipping/ShippingZone.php:
```diff
@@ -78,6 +78,79 @@ public function get_shipping_methods_for_country( string $country_code ): array
+		$rates = [];
+		foreach ( $methods as $method ) {
+			// We process the free shipping method separately.
+			if ( ! $method['enabled'] || self::METHOD_FREE === $method['id'] ) {
```
❓ 

What would happen if the user only has a free shipping method for a country? For example, imagine this scenario:

Spain: Only method available is Free Shipping
Portugal: Only method available is a Flat rate

When I tried to sync the rates I get only the rate for Portugal and Spain does not have any rate.

nima-karimi replied:
Oh yeah, that shouldn't be 😅  Fixed it in https://github.com/woocommerce/google-listings-and-ads/pull/1192/commits/ca42c16668e13fc286fe0a31f69bc6a4313da916

jorgemd24 reviewed (commented):
Thanks, @nima-karimi for this amazing work! 😊  I have checked the `shippingZone file` as I am getting familiar with the code and the project. I will continue checking the rest of the files during the next few days. 

See some comments that I left in the code.

Commit:
Check if free shipping method is enabled

Commit:
Remove unused method for checking valid methods

Commit:
Revert "Remove unused method for checking valid methods"

This reverts commit 7f72b04660228b1103795baa9a31ffb1f1994dde.

Commit:
Add method to check if shipping method is supported

Commit:
Return the free shipping rate if it's the only rate

Commit:
Sync minimum order value if the service rate is 0

If the shipping service is a free service, we also need to set the minimum order value because that might be the only available rate.

nima-karimi commented:
Thanks for the review, @jorgemd24 🙏 I've added some commits based on your comments. Ready for another review.

Commit:
Fix the rates output

Commit:
Remove tests for pickup method since we don't support it yet

Commit:
Remove all indications of pickup shipping method

We currently do not support the pickup shipping method.

jorgemd24 commented on src/Shipping/ShippingZone.php:
```diff
@@ -78,6 +77,89 @@ public function get_shipping_methods_for_country( string $country_code ): array
+		$methods = $this->get_shipping_methods_for_country( $country_code );
+
+		// If the only shipping method is the free shipping method, return it as the only rate.
+		if ( 1 === count( $methods ) && self::METHOD_FREE === $methods[0]['id'] ) {
+			$rate = [
+				'country'  => $country_code,
+				'method'   => $methods[0]['id'],
+				'currency' => $methods[0]['currency'],
+				'rate'     => 0,
+				'options'  => [],
+			];
+			if ( isset( $methods[0]['options']['min_amount'] ) ) {
+				$rate['options']['free_shipping_threshold'] = $methods[0]['options']['min_amount'];
+			}
+
+			return [ $rate ];
+		}
```
Thanks for adding this though I think the problem is not fixed yet, imagine this scenario where you have a country with:

- Free Shipping (enable)
- Flat rate (disabled)

It will not sync the free shipping because count(methods) = 2. 

What do you think if we filter all the disabled methods before we continue with the execution and we could also have a method to `parse_rates_contry` so we could re-use the code and it will be easier to test. 

```
$methods = $this->get_shipping_methods_for_country( $country_code );
$methods_enabled = filter only enable methods 

if ( 1 === count( $methods_enabled ) && self::METHOD_FREE === $methods_enabled[0]['id'] ) {
  return [$this->parse_rates_country( $methods_enabled[0] )]

}


```

nima-karimi replied:
Oh you're right. Didn't consider that case. I've added some commits including unit tests to prevent that issue.

jorgemd24 replied:
Thanks! it is working now 😃 

❓ 

Why do we not use the `parse_method` instead of doing it manually? 

```
			$rate = [
				'country'  => $country_code,
				'method'   => $methods[0]['id'],
				'currency' => $methods[0]['currency'],
				'rate'     => 0,
				'options'  => [],
			];
			if ( isset( $methods[0]['options']['min_amount'] ) ) {
				$rate['options']['free_shipping_threshold'] = $methods[0]['options']['min_amount'];
			}

```

nima-karimi replied:
The structure that we expect here is a bit different than what is returned by `parse_method`. This function returns the rates as expected by our API so the data structure is changed to be similar to how we store the rates in DB.

jorgemd24 replied:
Yes, you are right I got a bit confused there!  🙃  It is a different structure.

jorgemd24 commented on src/API/Site/Controllers/MerchantCenter/ShippingRateController.php:
```diff
@@ -317,28 +318,52 @@ protected function get_all_shipping_rates(): array {
+		$rates_output = [];
+		foreach ( $rates as $rate ) {
+			// We don't render the class rates because they are not used by the API.
+			if ( isset( $rate['options']['shipping_class_rates'] ) ) {
```
💅 

I think the `if conditional` could be remove and we just have ` unset( $rate['options']['shipping_class_rates'] ); `
```
			if ( isset( $rate['options']['shipping_class_rates'] ) ) {
				unset( $rate['options']['shipping_class_rates'] );
			}
```

nima-karimi replied:
Done. https://github.com/woocommerce/google-listings-and-ads/pull/1192/commits/aff1101e9655f8ae61ce0e4799117955f7747fc9

jorgemd24 replied:
Excellent thanks! 👍

jorgemd24 reviewed (commented):
Thanks, @nima-karimi for the improvements! 👍  

I left just some comments about `get_shipping_rates_for_country`. 

❔ 

I see in this PR some changes about the $target_audience however I do not see the relation with syncing the rates. How is it related?

Commit:
Return only enabled shipping methods

Commit:
Fix typo

Commit:
Add unit test for returning disabled shipping methods

Commit:
Add unit test for free shipping only scenario

Commit:
No need to check if array key is set before un-setting

nima-karimi commented:
> I see in this PR some changes about the $target_audience however I do not see the relation with syncing the rates. How is it related?

This is so that the `automatic` setting can work without any user input. It gets the list of target countries the user has previously set in the onboarding wizard and tries to fetch their shipping rates and sync them to Google.

nima-karimi commented:
Ready for another review 🙏

Commit:
Merge branch 'develop' into feature/add-automatic-setting-for-shipping

Commit:
Merge branch 'feature/refactor-shipping-rates-api' into feature/add-automatic-setting-for-shipping

jorgemd24 commented on src/Shipping/ShippingZone.php:
```diff
@@ -101,10 +183,14 @@ protected function parse_shipping_zones(): void {
 			foreach ( $methods as $method ) {
-				// Check if the method is supported and return its properties.
+				// Skip if the method is unsupported or disabled.
+				if ( ! $method->is_enabled() || ! self::is_shipping_method_supported( $method->id ) ) {
```
❓ 

It is a bit confusing for me that the method `parse_shipping_zones` is the one that is filtering the methods. 
I would expect that parse_shipping_zone would only parse the methods independently if it is enabled or if it is a shipping method supported.  What do you think?

nima-karimi replied:
I think it's okay to leave the filtering in there because it's parsing the shipping zones and not the methods. It's actually already filtering those that it can't support. The main task here is to go through the WooCommerce shipping zone settings and return the methods that we support in the format that we expect.

jorgemd24 replied:
Thanks for the clarification! 👍  

💅 

Just to avoid this confusion in the future, we could update the method description from: 

`Parses the WooCommerce shipping zones and maps them into the self::$methods_countries array.`

To (or something similar):

`Parses the WooCommerce shipping zones, maps them into the self::$methods_countries array and  filters the methods that are not enabled or supported`

jorgemd24 commented on src/API/Google/Settings.php:
```diff
@@ -51,17 +52,52 @@ public function sync_shipping() {
+			if ( isset( $options['free_shipping_threshold'] ) ) {
+				$minimum_order_value = (float) $options['free_shipping_threshold'];
+
+				if ( 0 !== $rate ) {
```
💅 

I think this could be $rate >= 0 just in case someone puts a negative value. I tried setting a negative value and Google throws an error.

nima-karimi replied:
This part only checks if the rate is free or not. I added another check on top for negative values.

jorgemd24 replied:
Perfect thanks!

jorgemd24 commented on src/API/Google/Settings.php:
```diff
@@ -51,17 +52,52 @@ public function sync_shipping() {
+	 *
+	 * @since x.x.x
+	 */
+	public function update_shipping_rates( array $rates ) {
```
💅 

It is missing the return type.

nima-karimi replied:
hmm. but it doesn't return anything

jorgemd24 replied:
I think the return type could be `void` as it doesn't return anything.

jorgemd24 commented on src/API/Google/Settings.php:
```diff
@@ -155,12 +202,33 @@ protected function get_times(): array {
+			$rates         = array_merge( $rates, $country_rates );
+		}
+
+		return $rates;
```
❓ 

When I try to sync the shipping rates for Austria I get the following error:

```
{
    "code": "gla_setting_sync_error",
    "message": "{\n  \"error\": {\n    \"code\": 400,\n    \"message\": \"[services[0].deliveryTime] Missing required service delivery time\",\n    \"errors\": [\n      {\n        \"message\": \"[services[0].deliveryTime] Missing required service delivery time\",\n        \"domain\": \"global\",\n        \"reason\": \"required\"\n      }\n    ]\n  }\n}\n",
    "data": {
        "status": 500,
        "message": "",
        "data": {
            "error": {
                "code": 400,
                "message": "[services[0].deliveryTime] Missing required service delivery time",
                "errors": [
                    {
                        "message": "[services[0].deliveryTime] Missing required service delivery time",
                        "domain": "global",
                        "reason": "required"
                    }
                ]
            }
        }
    }
}
```
I think Google requires deliveryTime for some countries?

nima-karimi replied:
deliveryTime is required but that should already be set if you go through the onboarding wizard. Can you edit your free-listing settings or go through onboarding once again and see if the issue will be resolved? It should be saved under the `gla_merchant_center` option in the `wp_options` table with the key `shipping_time`.

jorgemd24 replied:
Yes, the issue was that I chose "I have a more complex shipping setup and I cannot estimate flat shipping rates and times." in the wizard setup. Once I set the shipping time in the wizard went all good. 👍

jorgemd24 reviewed (commented):
Thanks @nima-karimi for the adjustments! 😃 

I left some notes in the code.

Commit:
Check for negative rates when syncing shipping settings

nima-karimi commented:
Left some comments.  Ready for the fourth review 😅

jorgemd24 reviewed (approved):
Thanks, @nima-karimi for addressing those changes! 💪 

What I have tested:

- ✅  /gla/mc/settings updates the shipping rate type.
- ✅  /gla/mc/settings/sync syncs the WC rates to Google MC.
- ✅  I have tested with free shipping, the minimum amount for free shipping, shipping classes and different countries.

LGTM 👍

Commit:
Clarify what the `parse_shipping_zones` method does in comment

nima-karimi commented:
Thanks so much for reviewing this PR multiple times 🙏

I've added a comment to clarify what the `parse_shipping_zones` method does in [comment](https://github.com/woocommerce/google-listings-and-ads/pull/1192/commits/dbd522c2e2046b2e4f1c34830143a81681516e8e)

> I think the return type could be void as it doesn't return anything.

We don't really have a rule for this, to be honest. We sometimes follow this pattern and sometimes don't. In the `Settings` class where this method is located, there are other methods that don't return anything but their return type is not set to void.
